markets/storageadapter/dealpublisher.go

```diff
@@ -434,12 +460,41 @@ func pieceCids(deals []market.ClientDealProposal) string {
 
 // filter out deals that have been cancelled
 func (p *DealPublisher) filterCancelledDeals() {
-	filtered := p.pending[:0]
-	for _, pd := range p.pending {
+	filtered := make(map[cid.Cid]*pendingDeal)
+	for c, pd := range p.pending {
 		if pd.ctx.Err() != nil {
 			continue
 		}
-		filtered = append(filtered, pd)
+		filtered[c] = pd
 	}
 	p.pending = filtered
```

Comment:
Rather than creating a new map, I think you can just delete from the existing map